File: FateGrandAutomata.Core/Modules/Scaling.cs
using System;
using CoreAutomata;

namespace FateGrandAutomata
{
    public static class Scaling
    {
        static (bool ByWidth, double Rate) DecideScaleMethod(int OriginalWidth, int OriginalHeight, int DesiredWidth, int DesiredHeight)
        {
            var rateToScaleByWidth = DesiredWidth / (double) OriginalWidth;
            var rateToScaleByHeight = DesiredHeight / (double) OriginalHeight;

            return rateToScaleByWidth <= rateToScaleByHeight
                ? (true, rateToScaleByWidth)
                : (false, rateToScaleByHeight);
        }

        static (int Width, int Height) Scale(int OriginalWidth, int OriginalHeight, double Rate)
        {
            var w = (int) Math.Round(OriginalWidth * Rate);
            var h = (int) Math.Round(OriginalHeight * Rate);

            return (w, h);
        }

        static int CalculateBorderThickness(int Outer, int Inner)
        {
            var size = Math.Abs(Outer - Inner);

            return (int) Math.Round(size / 2.0);
        }

        static Region CalculateGameAreaWithoutBorders(int ScriptWidth,
            int ScriptHeight,
            int ScreenWidth,
            int ScreenHeight,
            double ScaleRate)
        {
            var (scaledW, scaledH) = Scale(ScriptWidth, ScriptHeight, ScaleRate);

            return new Region(
                CalculateBorderThickness(ScreenWidth, scaledW), // Offset (X)
                CalculateBorderThickness(ScreenHeight, scaledH), // Offset (Y)
                scaledW, // Game Width (without borders)
                scaledH // Game Height (without borders)
            );
        }

        static Region ApplyNotchOffset(Region Region, int NotchOffset)
        {
            Region.X += NotchOffset;

            return Region;
        }

        static void ApplyAspectRatioFix(int ScriptWidth, int ScriptHeight, int ImageWidth, int ImageHeight)
        {
            var gameWithBorders = GameAreaManager.GameArea;
            var (scaleByWidth, scaleRate) = DecideScaleMethod(ScriptWidth, ScriptHeight,
                gameWithBorders.W, gameWithBorders.H);
            var gameWithoutBorders = CalculateGameAreaWithoutBorders(ScriptWidth, ScriptHeight,
                gameWithBorders.W, gameWithBorders.H,
                scaleRate);
            var gameWithoutBordersAndNotch = ApplyNotchOffset(gameWithoutBorders, gameWithBorders.X);

            GameAreaManager.GameArea = gameWithoutBordersAndNotch;

            if (scaleByWidth)
            {
                GameAreaManager.ScriptDimension = new CompareSettings(true, ScriptWidth);
                GameAreaManager.CompareDimension = new CompareSettings(true, ImageWidth);
            }
            else
            {
                GameAreaManager.ScriptDimension = new CompareSettings(false, ScriptHeight);
                GameAreaManager.CompareDimension = new CompareSettings(false, ImageHeight);
            }
        }

        public static void Init()
        {
            GameAreaManager.Reset();

            // Set only ONCE
            ApplyAspectRatioFix(Game.ScriptWidth,
                Game.ScriptHeight,
                Game.ImageWidth,
                Game.ImageHeight);
        }
    }
}

File: scripts/src/main/java/com/mathewsachin/fategrandautomata/scripts/modules/Scaling.kt
package com.mathewsachin.fategrandautomata.scripts.modules

import com.mathewsachin.libautomata.CompareBy
import com.mathewsachin.libautomata.GameAreaManager
import com.mathewsachin.libautomata.Region
import com.mathewsachin.libautomata.Size
import com.mathewsachin.libautomata.dagger.ScriptScope
import javax.inject.Inject
import kotlin.math.abs
import kotlin.math.roundToInt

private sealed class ScaleBy(val rate: Double) {
    class Width(rate: Double) : ScaleBy(rate)
    class Height(rate: Double) : ScaleBy(rate)
}

private fun decideScaleMethod(OriginalSize: Size, DesiredSize: Size): ScaleBy {
    val rateToScaleByWidth = DesiredSize.Width / OriginalSize.Width.toDouble()
    val rateToScaleByHeight = DesiredSize.Height / OriginalSize.Height.toDouble()

    return if (rateToScaleByWidth <= rateToScaleByHeight)
        ScaleBy.Width(rateToScaleByWidth)
    else ScaleBy.Height(rateToScaleByHeight)
}

private fun calculateBorderThickness(Outer: Int, Inner: Int): Int {
    val size = abs(Outer - Inner)

    return (size / 2.0).roundToInt()
}

private fun calculateGameAreaWithoutBorders(
    ScriptSize: Size,
    ScreenSize: Size,
    ScaleRate: Double
): Region {
    val scaledScriptSize = ScriptSize * ScaleRate

    return Region(
        calculateBorderThickness(
            ScreenSize.Width,
            scaledScriptSize.Width
        ), // Offset(X)
        calculateBorderThickness(
            ScreenSize.Height,
            scaledScriptSize.Height
        ), // Offset(Y)
        scaledScriptSize.Width, // Game Width (without borders)
        scaledScriptSize.Height // Game Height (without borders)
    )
}

@ScriptScope
class Scaling @Inject constructor(val gameAreaManager: GameAreaManager, val game: Game) {
    private fun applyAspectRatioFix(ScriptSize: Size, ImageSize: Size) {
        val gameWithBorders = gameAreaManager.gameArea
        val scaleBy = decideScaleMethod(
            ScriptSize,
            gameWithBorders.size
        )
        val gameWithoutBorders =
            calculateGameAreaWithoutBorders(
                ScriptSize,
                gameWithBorders.size,
                scaleBy.rate
            )

        gameAreaManager.gameArea = gameWithoutBorders

        gameAreaManager.scriptDimension = when (scaleBy) {
            is ScaleBy.Width -> CompareBy.Width(ScriptSize.Width)
            is ScaleBy.Height -> CompareBy.Height(ScriptSize.Height)
        }

        gameAreaManager.compareDimension = when (scaleBy) {
            is ScaleBy.Width -> CompareBy.Width(ImageSize.Width)
            is ScaleBy.Height -> CompareBy.Height(ImageSize.Height)
        }
    }

    fun init() {
        applyAspectRatioFix(
            Game.scriptSize,
            Game.imageSize
        )
    }
}